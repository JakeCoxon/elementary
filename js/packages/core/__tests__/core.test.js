import {
  createNode,
  renderWithDelegate,
  stepGarbageCollector,
} from '..';


class TestRenderer {
  constructor(config) {
    this.nodeMap = new Map();
    this.memoMap = new Map();
    this.batch = [];
    this.roots = [];
    this.config = Object.assign({
      logCreateNode: true,
      logDeleteNode: true,
      logAppendChild: true,
      logSetProperty: true,
      logActivateRoots: true,
      logCommitUpdates: true,
    }, config);
  }

  getNodeMap() {
    return this.nodeMap;
  }

  getMemoMap() {
    return this.memoMap;
  }

  getRenderContext() {
    return {};
  }

  getActiveRoots() {
    return this.roots;
  }

  getTerminalGeneration() {
    return 4;
  }

  getBatch() {
    return this.batch;
  }

  clearBatch() {
    this.batch = [];
  }

  createNode(...args) {
    if (this.config.logCreateNode) {
      this.batch.push(["createNode", ...args]);
    }
  }

  deleteNode(...args) {
    if (this.config.logDeleteNode) {
      this.batch.push(["deleteNode", ...args]);
    }
  }

  appendChild(...args) {
    if (this.config.logAppendChild) {
      this.batch.push(["appendChild", ...args]);
    }
  }

  setProperty(...args) {
    if (this.config.logSetProperty) {
      this.batch.push(["setProperty", ...args]);
    }
  }

  activateRoots(...args) {
    if (this.config.logActivateRoots) {
      this.batch.push(["activateRoots", ...args]);
    }
  }

  commitUpdates(...args) {
    if (this.config.logCommitUpdates) {
      this.batch.push(["commitUpdates", ...args]);
    }
  }

  render(...args) {
    this.roots = renderWithDelegate(this, args);
  }

  gc() {
    stepGarbageCollector(this);
  }
}

test('the basics', function() {
  let tr = new TestRenderer();

  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        createNode("const", {value: 2 * Math.PI}, []),
        createNode("phasor", {}, [
          createNode("const", {key: 'fq', value: 440}, [])
        ])
      ])
    ])
  );

  expect(tr.getBatch()).toMatchSnapshot();
});

test('numeric literals', function() {
  let tr = new TestRenderer();

  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [440])
      ])
    ])
  );

  expect(tr.getBatch()).toMatchSnapshot();
});

test('distinguish by props', function() {
  let tr = new TestRenderer();

  const renderVoice = (path, seq) => {
    return (
      createNode("sample", {path}, [
        createNode("seq", {seq}, [
          createNode("le", {}, [
            createNode("phasor", {}, [
              createNode("const", {value: 2}, [])
            ]),
            createNode("const", {value: 0.5}, [])
          ]),
        ]),
      ])
    );
  };

  tr.render(
    renderVoice('test/path.wav', [0, 0, 1]),
    renderVoice('test/path.wav', [0, 1, 0]),
  );

  // Here we haven't specified any keys, so the renderer should do a lot of
  // structural sharing. However, because we're rendering a different sequence,
  // we expect to see the pulse train (starting at the "le" node) shared, but
  // then see two different root nodes, two different sample nodes, and two
  // different seq nodes.
  expect(tr.getBatch()).toMatchSnapshot();
});

test('multi-channel basics', function() {
  const tr = new TestRenderer();

  const monoProcess = (
    createNode("sin", {}, [
      createNode("mul", {}, [
        createNode("const", {value: 2 * Math.PI}, []),
        createNode("phasor", {}, [
          createNode("const", {key: 'fq', value: 440}, [])
        ])
      ])
    ])
  );

  // Run the same thing in the left and the right channel
  tr.render(monoProcess, monoProcess);

  // We should see structural sharing except for the root nodes
  expect(tr.getBatch()).toMatchSnapshot();
});

test('simple sharing', function() {
  let tr = new TestRenderer();

  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        createNode("const", {value: 2 * Math.PI}, []),
        createNode("phasor", {}, [
          createNode("const", {key: 'fq', value: 440}, [])
        ])
      ])
    ])
  );

  // Clear the call stack
  tr.clearBatch();

  // Second render inserts a tanh at the top, we should find this and
  // share all of the child nodes during this render pass
  tr.render(
    createNode("tanh", {}, [
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: 'fq', value: 440}, [])
          ])
        ])
      ])
    ])
  );

  expect(tr.getBatch()).toMatchSnapshot();
});

test('distinguished subtrees by key', function() {
  let tr = new TestRenderer();

  const voices = [
    {key: 'fq1', freq: 440},
    {key: 'fq2', freq: 440},
    {key: 'fq3', freq: 440},
    {key: 'fq4', freq: 440},
  ];

  const renderVoice = (voice) => {
    return (
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: voice.key, value: voice.freq}, [])
          ])
        ])
      ])
    );
  };

  tr.render(createNode("add", {}, voices.map(renderVoice)))

  expect(tr.getBatch()).toMatchSnapshot();
});

test('structural equality with value change', function() {
  const tr = new TestRenderer();

  const voices = [
    {key: 'fq1', freq: 440},
    {key: 'fq2', freq: 440},
    {key: 'fq3', freq: 440},
    {key: 'fq4', freq: 440},
  ];

  const renderVoice = (voice) => {
    return (
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: voice.key, value: voice.freq}, [])
          ])
        ])
      ])
    );
  };

  tr.render(createNode("add", {}, voices.map(renderVoice)))

  // Clear the call stack
  tr.clearBatch();

  // Change one of the keyed values, should see structural equality
  // with no new nodes created
  voices[0].freq = 441;

  tr.render(createNode("add", {}, voices.map(renderVoice)))

  expect(tr.getBatch()).toMatchSnapshot();
});

// Testing here to ensure that the root node activation works as expected.
test('switch and switch back', function() {
  const tr = new TestRenderer();

  const renderVoice = (voice) => {
    return (
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: voice.key, value: voice.freq}, [])
          ])
        ])
      ])
    );
  };

  // We're going to render three times with graphs A, B, A. We want to see on
  // the third render that almost nothing was done because we still have the nodes
  // around that we need, i.e. they haven't been garbage collected, but we will need
  // to see that the correct root nodes are in place.
  tr.render(renderVoice({key: 'hi', freq: 440}));
  tr.render(renderVoice({key: 'bye', freq: 880}));

  tr.clearBatch();

  tr.render(renderVoice({key: 'hi', freq: 440}));
  expect(tr.getBatch()).toMatchSnapshot();
});

test('composite with keyed children', function() {
  const tr = new TestRenderer();
  const composite = ({context, props, children}) => {
    return createNode("add", {}, children);
  };

  tr.render(createNode(composite, {}, [
    createNode("const", {key: 'g', value: 0}, [])
  ]));

  expect(tr.getBatch()).toMatchSnapshot();
  tr.clearBatch();

  tr.render(createNode(composite, {}, [
    createNode("const", {key: 'g', value: 1}, [])
  ]));

  expect(tr.getBatch()).toMatchSnapshot();
});

test('shared reference to composite node', function() {
  let tr = new TestRenderer();
  let calls = 0;

  let train = ({props, children: [rate]}) => {
    calls++;

    return (
      createNode("le", {}, [
        createNode("phasor", {}, [rate]),
        0.5,
      ])
    );
  };

  // We expect `train` to be called once in resolving the composite node, and for
  // the latter two instances in our stereo graph to be resolved via the cache
  let t = createNode(train, {}, [5]);

  tr.render(
    createNode("seq", {seq: [1, 2, 3]}, [t, t]),
    createNode("seq", {seq: [1, 2, 3]}, [t]),
  );

  expect(calls).toBe(1);

  // Now if we make multiple composite node instances each referencing the same composite
  // function, we have to unroll each of them with their associated inputs. Due to the v2.0.1
  // change which disables memoization we no longer attempt to prevent that second call by memoization
  calls = 0;

  let u = createNode(train, {}, [5]);
  let v = createNode(train, {}, [5]);

  tr.render(
    createNode("seq", {seq: [1, 2, 3]}, [u, v]),
    createNode("seq", {seq: [1, 2, 3]}, [v]),
  );

  expect(calls).toBe(2);
});

test('garbage collection', function() {
  let tr = new TestRenderer();

  // First, render a sine tone with `sin`
  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [440])
      ])
    ])
  );

  // Next we're going to render a sine tone using cosine. We'll expect to
  // see most of the prior structure preserved.
  tr.clearBatch();
  tr.gc();

  tr.render(
    createNode("cos", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [440])
      ])
    ])
  );

  expect(tr.getBatch()).toMatchSnapshot();

  // Now if we step the garbage collector enough we should see the sine node
  // and its parent root get cleaned up now that they're no longer referenced
  // in the active tree
  for (let i = 0; i < tr.getTerminalGeneration() - 1; ++i) {
    tr.clearBatch();
    tr.gc();
  }

  expect(tr.getBatch()).toMatchSnapshot();
});
