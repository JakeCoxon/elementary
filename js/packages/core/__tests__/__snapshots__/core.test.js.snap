// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composite with keyed children 1`] = `
Array [
  Array [
    "createNode",
    1071548371,
    "const",
  ],
  Array [
    "setProperty",
    1071548371,
    "key",
    "g",
  ],
  Array [
    "setProperty",
    1071548371,
    "value",
    0,
  ],
  Array [
    "createNode",
    985951761,
    "add",
  ],
  Array [
    "appendChild",
    985951761,
    1071548371,
  ],
  Array [
    "createNode",
    1374919193,
    "root",
  ],
  Array [
    "setProperty",
    1374919193,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    1374919193,
    985951761,
  ],
  Array [
    "activateRoots",
    Array [
      1374919193,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`composite with keyed children 2`] = `
Array [
  Array [
    "setProperty",
    1071548371,
    "value",
    1,
  ],
  Array [
    "activateRoots",
    Array [
      1374919193,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`distinguish by props 1`] = `
Array [
  Array [
    "createNode",
    1557538697,
    "const",
  ],
  Array [
    "setProperty",
    1557538697,
    "value",
    2,
  ],
  Array [
    "createNode",
    1616670699,
    "phasor",
  ],
  Array [
    "appendChild",
    1616670699,
    1557538697,
  ],
  Array [
    "createNode",
    1181055054,
    "const",
  ],
  Array [
    "setProperty",
    1181055054,
    "value",
    0.5,
  ],
  Array [
    "createNode",
    1049815855,
    "le",
  ],
  Array [
    "appendChild",
    1049815855,
    1616670699,
  ],
  Array [
    "appendChild",
    1049815855,
    1181055054,
  ],
  Array [
    "createNode",
    1089078935,
    "seq",
  ],
  Array [
    "setProperty",
    1089078935,
    "seq",
    Array [
      0,
      0,
      1,
    ],
  ],
  Array [
    "appendChild",
    1089078935,
    1049815855,
  ],
  Array [
    "createNode",
    1440666767,
    "sample",
  ],
  Array [
    "setProperty",
    1440666767,
    "path",
    "test/path.wav",
  ],
  Array [
    "appendChild",
    1440666767,
    1089078935,
  ],
  Array [
    "createNode",
    1840069255,
    "root",
  ],
  Array [
    "setProperty",
    1840069255,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    1840069255,
    1440666767,
  ],
  Array [
    "createNode",
    493667567,
    "seq",
  ],
  Array [
    "setProperty",
    493667567,
    "seq",
    Array [
      0,
      1,
      0,
    ],
  ],
  Array [
    "appendChild",
    493667567,
    1049815855,
  ],
  Array [
    "createNode",
    1741781271,
    "sample",
  ],
  Array [
    "setProperty",
    1741781271,
    "path",
    "test/path.wav",
  ],
  Array [
    "appendChild",
    1741781271,
    493667567,
  ],
  Array [
    "createNode",
    383156298,
    "root",
  ],
  Array [
    "setProperty",
    383156298,
    "channel",
    1,
  ],
  Array [
    "appendChild",
    383156298,
    1741781271,
  ],
  Array [
    "activateRoots",
    Array [
      1840069255,
      383156298,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`distinguished subtrees by key 1`] = `
Array [
  Array [
    "createNode",
    1097289128,
    "const",
  ],
  Array [
    "setProperty",
    1097289128,
    "value",
    6.283185307179586,
  ],
  Array [
    "createNode",
    1427558172,
    "const",
  ],
  Array [
    "setProperty",
    1427558172,
    "key",
    "fq1",
  ],
  Array [
    "setProperty",
    1427558172,
    "value",
    440,
  ],
  Array [
    "createNode",
    453812052,
    "phasor",
  ],
  Array [
    "appendChild",
    453812052,
    1427558172,
  ],
  Array [
    "createNode",
    453550641,
    "mul",
  ],
  Array [
    "appendChild",
    453550641,
    1097289128,
  ],
  Array [
    "appendChild",
    453550641,
    453812052,
  ],
  Array [
    "createNode",
    1074324112,
    "sin",
  ],
  Array [
    "appendChild",
    1074324112,
    453550641,
  ],
  Array [
    "createNode",
    1193164375,
    "const",
  ],
  Array [
    "setProperty",
    1193164375,
    "key",
    "fq2",
  ],
  Array [
    "setProperty",
    1193164375,
    "value",
    440,
  ],
  Array [
    "createNode",
    197128117,
    "phasor",
  ],
  Array [
    "appendChild",
    197128117,
    1193164375,
  ],
  Array [
    "createNode",
    364394878,
    "mul",
  ],
  Array [
    "appendChild",
    364394878,
    1097289128,
  ],
  Array [
    "appendChild",
    364394878,
    197128117,
  ],
  Array [
    "createNode",
    787548653,
    "sin",
  ],
  Array [
    "appendChild",
    787548653,
    364394878,
  ],
  Array [
    "createNode",
    555467758,
    "const",
  ],
  Array [
    "setProperty",
    555467758,
    "key",
    "fq3",
  ],
  Array [
    "setProperty",
    555467758,
    "value",
    440,
  ],
  Array [
    "createNode",
    1492179306,
    "phasor",
  ],
  Array [
    "appendChild",
    1492179306,
    555467758,
  ],
  Array [
    "createNode",
    220712335,
    "mul",
  ],
  Array [
    "appendChild",
    220712335,
    1097289128,
  ],
  Array [
    "appendChild",
    220712335,
    1492179306,
  ],
  Array [
    "createNode",
    1512310154,
    "sin",
  ],
  Array [
    "appendChild",
    1512310154,
    220712335,
  ],
  Array [
    "createNode",
    1661951969,
    "const",
  ],
  Array [
    "setProperty",
    1661951969,
    "key",
    "fq4",
  ],
  Array [
    "setProperty",
    1661951969,
    "value",
    440,
  ],
  Array [
    "createNode",
    79046899,
    "phasor",
  ],
  Array [
    "appendChild",
    79046899,
    1661951969,
  ],
  Array [
    "createNode",
    1376555748,
    "mul",
  ],
  Array [
    "appendChild",
    1376555748,
    1097289128,
  ],
  Array [
    "appendChild",
    1376555748,
    79046899,
  ],
  Array [
    "createNode",
    866439039,
    "sin",
  ],
  Array [
    "appendChild",
    866439039,
    1376555748,
  ],
  Array [
    "createNode",
    1808164672,
    "add",
  ],
  Array [
    "appendChild",
    1808164672,
    1074324112,
  ],
  Array [
    "appendChild",
    1808164672,
    787548653,
  ],
  Array [
    "appendChild",
    1808164672,
    1512310154,
  ],
  Array [
    "appendChild",
    1808164672,
    866439039,
  ],
  Array [
    "createNode",
    19209078,
    "root",
  ],
  Array [
    "setProperty",
    19209078,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    19209078,
    1808164672,
  ],
  Array [
    "activateRoots",
    Array [
      19209078,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`garbage collection 1`] = `
Array [
  Array [
    "createNode",
    1511449914,
    "cos",
  ],
  Array [
    "appendChild",
    1511449914,
    1913760094,
  ],
  Array [
    "createNode",
    1305352856,
    "root",
  ],
  Array [
    "setProperty",
    1305352856,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    1305352856,
    1511449914,
  ],
  Array [
    "activateRoots",
    Array [
      1305352856,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`garbage collection 2`] = `
Array [
  Array [
    "deleteNode",
    520300429,
  ],
  Array [
    "deleteNode",
    1423955117,
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`multi-channel basics 1`] = `
Array [
  Array [
    "createNode",
    1097289128,
    "const",
  ],
  Array [
    "setProperty",
    1097289128,
    "value",
    6.283185307179586,
  ],
  Array [
    "createNode",
    1709345783,
    "const",
  ],
  Array [
    "setProperty",
    1709345783,
    "key",
    "fq",
  ],
  Array [
    "setProperty",
    1709345783,
    "value",
    440,
  ],
  Array [
    "createNode",
    1952747413,
    "phasor",
  ],
  Array [
    "appendChild",
    1952747413,
    1709345783,
  ],
  Array [
    "createNode",
    1689717726,
    "mul",
  ],
  Array [
    "appendChild",
    1689717726,
    1097289128,
  ],
  Array [
    "appendChild",
    1689717726,
    1952747413,
  ],
  Array [
    "createNode",
    872803085,
    "sin",
  ],
  Array [
    "appendChild",
    872803085,
    1689717726,
  ],
  Array [
    "createNode",
    805437229,
    "root",
  ],
  Array [
    "setProperty",
    805437229,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    805437229,
    872803085,
  ],
  Array [
    "createNode",
    1850081212,
    "root",
  ],
  Array [
    "setProperty",
    1850081212,
    "channel",
    1,
  ],
  Array [
    "appendChild",
    1850081212,
    872803085,
  ],
  Array [
    "activateRoots",
    Array [
      805437229,
      1850081212,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`numeric literals 1`] = `
Array [
  Array [
    "createNode",
    1097289128,
    "const",
  ],
  Array [
    "setProperty",
    1097289128,
    "value",
    6.283185307179586,
  ],
  Array [
    "createNode",
    198353271,
    "const",
  ],
  Array [
    "setProperty",
    198353271,
    "value",
    440,
  ],
  Array [
    "createNode",
    1406216981,
    "phasor",
  ],
  Array [
    "appendChild",
    1406216981,
    198353271,
  ],
  Array [
    "createNode",
    1913760094,
    "mul",
  ],
  Array [
    "appendChild",
    1913760094,
    1097289128,
  ],
  Array [
    "appendChild",
    1913760094,
    1406216981,
  ],
  Array [
    "createNode",
    520300429,
    "sin",
  ],
  Array [
    "appendChild",
    520300429,
    1913760094,
  ],
  Array [
    "createNode",
    1423955117,
    "root",
  ],
  Array [
    "setProperty",
    1423955117,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    1423955117,
    520300429,
  ],
  Array [
    "activateRoots",
    Array [
      1423955117,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`simple sharing 1`] = `
Array [
  Array [
    "createNode",
    23247763,
    "tanh",
  ],
  Array [
    "appendChild",
    23247763,
    872803085,
  ],
  Array [
    "createNode",
    188983667,
    "root",
  ],
  Array [
    "setProperty",
    188983667,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    188983667,
    23247763,
  ],
  Array [
    "activateRoots",
    Array [
      188983667,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`structural equality with value change 1`] = `
Array [
  Array [
    "setProperty",
    1427558172,
    "value",
    441,
  ],
  Array [
    "activateRoots",
    Array [
      19209078,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`switch and switch back 1`] = `
Array [
  Array [
    "activateRoots",
    Array [
      70054491,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;

exports[`the basics 1`] = `
Array [
  Array [
    "createNode",
    1097289128,
    "const",
  ],
  Array [
    "setProperty",
    1097289128,
    "value",
    6.283185307179586,
  ],
  Array [
    "createNode",
    1709345783,
    "const",
  ],
  Array [
    "setProperty",
    1709345783,
    "key",
    "fq",
  ],
  Array [
    "setProperty",
    1709345783,
    "value",
    440,
  ],
  Array [
    "createNode",
    1952747413,
    "phasor",
  ],
  Array [
    "appendChild",
    1952747413,
    1709345783,
  ],
  Array [
    "createNode",
    1689717726,
    "mul",
  ],
  Array [
    "appendChild",
    1689717726,
    1097289128,
  ],
  Array [
    "appendChild",
    1689717726,
    1952747413,
  ],
  Array [
    "createNode",
    872803085,
    "sin",
  ],
  Array [
    "appendChild",
    872803085,
    1689717726,
  ],
  Array [
    "createNode",
    805437229,
    "root",
  ],
  Array [
    "setProperty",
    805437229,
    "channel",
    0,
  ],
  Array [
    "appendChild",
    805437229,
    872803085,
  ],
  Array [
    "activateRoots",
    Array [
      805437229,
    ],
  ],
  Array [
    "commitUpdates",
  ],
]
`;
